pkg/controllers/mpi_job_controller.go

```diff
@@ -0,0 +1,997 @@
+	// can attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+	if err != nil {
+		return nil, err
+	}
+	// If the launcher ServiceAccount is not controlled by this MPIJob resource, we
+	// should log a warning to the event recorder and return.
+	if !metav1.IsControlledBy(sa, mpiJob) {
+		msg := fmt.Sprintf(MessageResourceExists, sa.Name)
+		c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
+		return nil, fmt.Errorf(msg)
+	}
+
+	return sa, nil
+}
+
+// getLauncherRole gets the launcher Role controlled by this MPIJob.
+func (c *MPIJobController) getLauncherRole(mpiJob *kubeflow.MPIJob, workerReplicas int) (*rbacv1.Role, error) {
+	role, err := c.roleLister.Roles(mpiJob.Namespace).Get(mpiJob.Name + launcherSuffix)
+	// If the Role doesn't exist, we'll create it.
+	if errors.IsNotFound(err) {
+		role, err = c.kubeClient.RbacV1().Roles(mpiJob.Namespace).Create(newLauncherRole(mpiJob, workerReplicas))
+	}
+	// If an error occurs during Get/Create, we'll requeue the item so we
+	// can attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+	if err != nil {
+		return nil, err
+	}
+	// If the launcher Role is not controlled by this MPIJob resource, we
+	// should log a warning to the event recorder and return.
+	if !metav1.IsControlledBy(role, mpiJob) {
+		msg := fmt.Sprintf(MessageResourceExists, role.Name)
+		c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
+		return nil, fmt.Errorf(msg)
+	}
+
+	return role, nil
+}
+
+// getLauncherRoleBinding gets the launcher RoleBinding controlled by this MPIJob.
+func (c *MPIJobController) getLauncherRoleBinding(mpiJob *kubeflow.MPIJob) (*rbacv1.RoleBinding, error) {
+	rb, err := c.roleBindingLister.RoleBindings(mpiJob.Namespace).Get(mpiJob.Name + launcherSuffix)
+	// If the RoleBinding doesn't exist, we'll create it.
+	if errors.IsNotFound(err) {
+		rb, err = c.kubeClient.RbacV1().RoleBindings(mpiJob.Namespace).Create(newLauncherRoleBinding(mpiJob))
+	}
+	// If an error occurs during Get/Create, we'll requeue the item so we
+	// can attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+	if err != nil {
+		return nil, err
+	}
+	// If the launcher RoleBinding is not controlled by this MPIJob resource, we
+	// should log a warning to the event recorder and return.
+	if !metav1.IsControlledBy(rb, mpiJob) {
+		msg := fmt.Sprintf(MessageResourceExists, rb.Name)
+		c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
+		return nil, fmt.Errorf(msg)
+	}
+
+	return rb, nil
+}
+
+// getWorkerStatefulSet gets the worker StatefulSet controlled by this MPIJob.
+func (c *MPIJobController) getWorkerStatefulSet(mpiJob *kubeflow.MPIJob, workerReplicas int) (*appsv1.StatefulSet, error) {
```

Comment:
I think the name here does not reflect what it does. It also may create new statefulsets, IMO.

Reply:
OK, renamed these to `getOrCreate...`. Hope that's clearer.